### DISCUSSION TITLE
[FEAT] Access control provider - bulk check

### DISCUSSION THREAD

**Discussion by Mikilll94** [refinedev/refine#3964] [Ideas]
### Is your feature request related to a problem? Please describe.

I have noticed that `can` method from access control provider is executed for every row. This can lead to terrible performance when i.e. you need to make an http call to check permissions and have about 1000 rows. This will result in 1000 http calls.

Steps To Reproduce

1. Open this CodeSandbox.
https://codesandbox.io/s/sharp-snowflake-yxc8q9?file=/src/App.tsx
2. Log in
3. Look at the console. You will see that ACCESS CONTROL PROVIDER is printed 53 times.

### Describe the thing to improve

`can` method from access control provider should be executed only once.
Screenshot

**Comment by BatuhanW** [comment]
Hello @Mikilll94, thanks for the report.

By design, authorization usually goes through per record, not just resource and action. So we are covering these use-cases by providing this feature.

Also usually authorization servers are deployed together ideally in the same network so these requests are fast. For example, as a sidecar in kubernetes.

We are supporting different types of access controls like RBAC, Scope based.

As @omeraplak mentioned earlier [here](https://github.com/refinedev/refine/issues/3945#issuecomment-1478061383) you can disable this feature for buttons individually.

Since rendering 1000 rows in a page is a poor design and we can't possible handle %100 of the cases, we'll keep current approach here.

However, if you have some suggestion/API Designs to handle your case, we can discuss and eventually implement it.

**Comment by Mikilll94** [comment]
@BatuhanW
Sometimes the server does not support server-side pagination. In this case I fetch 1000 records from the server and handle pagination client-side. In this case 1000 http requests are made.

**Comment by KOCNS1** [comment]
> In this case 1000 http requests are made.

Just out of curiosity, how do you make 1000 requests to fetch 1000 records 🧐 and not just a single call that responds with the 1000 records?

**Comment by BatuhanW** [comment]
@KOCNS1, AFAIK @Mikilll94 makes a single query which returns 1000 records, next 1000 requests are made by refine access control provider for each record.

**Comment by BatuhanW** [comment]
> @BatuhanW Sometimes the server does not support server-side pagination. In this case I fetch 1000 records from the server and handle pagination client-side. In this case 1000 http requests are made.

I see, if you have some solution, API Design in mind, please share with us! We are happy to discuss and implement it.
